File: iOS/Kalorie/Core/Models/FoodItemSubmissionModel.swift
//
//  FoodItemSubmissionModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

enum FoodItemSubmissionStatus: String, Codable {
    case pending
    case rejected
}

struct FoodItemSubmissionDomain {

    // MARK: - Properties

    let id: String
    let barcode: String?
    let submittedBy: String
    let status: FoodItemSubmissionStatus
    let submittedAt: Date
    let rejectReason: String?
    let item: FoodItemDomain
}

enum FoodItemSubmissionError: Error {
    case invalidCode
    case invalidName
    case invalidCalories
    case invalidWeight
    case invalidPortion(FoodPortionError)
    case itemAlreadyExists

    init(_ validationError: FoodItemValidationError) {
        self = validationError.mapped(
            invalidCode: .invalidCode,
            invalidName: .invalidName,
            invalidCalories: .invalidCalories,
            invalidWeight: .invalidWeight
        ) { .invalidPortion($0) }
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/models/FoodItemSubmissionModel.kt
package antoni.kalorie.core.models

import java.time.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class FoodItemSubmissionStatus(val wireValue: String) {
    @SerialName("pending") PENDING("pending"),
    @SerialName("rejected") REJECTED("rejected"),
}

data class FoodItemSubmissionDomain(
    val id: String,
    val barcode: String?,
    val submittedBy: String,
    val status: FoodItemSubmissionStatus,
    val submittedAt: Instant,
    val rejectReason: String?,
    val item: FoodItemDomain,
)

sealed class FoodItemSubmissionError : Exception() {
    data object InvalidCode : FoodItemSubmissionError()
    data object InvalidName : FoodItemSubmissionError()
    data object InvalidCalories : FoodItemSubmissionError()
    data object InvalidWeight : FoodItemSubmissionError()
    data class InvalidPortion(val error: FoodPortionError) : FoodItemSubmissionError()
    data object ItemAlreadyExists : FoodItemSubmissionError()

    // MARK: - Functions

    companion object {
        fun from(validationError: FoodItemValidationError): FoodItemSubmissionError = validationError.mapped(
            invalidCode = InvalidCode,
            invalidName = InvalidName,
            invalidCalories = InvalidCalories,
            invalidWeight = InvalidWeight,
        ) { InvalidPortion(it) }
    }
}
